Common/src/linear_algebra/CSysSolve.cpp

```diff
@@ -916,64 +920,82 @@ unsigned long CSysSolve<ScalarType>::Solve(CSysMatrix<ScalarType> & Jacobian, co
     Residual = residual;
     Iterations = IterLinSol;
   }
+  END_SU2_OMP_MASTER
 
   HandleTemporariesOut(LinSysSol);
 
   delete precond;
 
   if(TapeActive) {
 
-    bool RequiresTranspose = !mesh_deform; // jacobian is symmetric
+    /*--- To keep the behavior of SU2_DOT, but not strictly required since jacobian is symmetric(?). ---*/
+    const bool RequiresTranspose = !mesh_deform || (config->GetKind_SU2() == SU2_DOT);
+
     if (!mesh_deform) KindPrecond = config->GetKind_DiscAdj_Linear_Prec();
     else              KindPrecond = config->GetKind_Deform_Linear_Solver_Prec();
 
-    /*--- Start recording if it was stopped for the linear solver ---*/
-
-    AD::StartRecording();
-
-    AD::SetExtFuncOut(&LinSysSol[0], (int)LinSysSol.GetLocSize());
-
-#ifdef CODI_REVERSE_TYPE
-    AD::FuncHelper->addUserData(&LinSysRes);
-    AD::FuncHelper->addUserData(&LinSysSol);
-    AD::FuncHelper->addUserData(&Jacobian);
-    AD::FuncHelper->addUserData(geometry);
-    AD::FuncHelper->addUserData(config);
-    AD::FuncHelper->addUserData(this);
-    AD::FuncHelper->addToTape(CSysSolve_b<ScalarType>::Solve_b);
-#endif
-
     /*--- Build preconditioner for the transposed Jacobian ---*/
 
+    if (RequiresTranspose) Jacobian.TransposeInPlace();
+
     switch(KindPrecond) {
       case ILU:
-        Jacobian.BuildILUPreconditioner(RequiresTranspose);
+        if (RequiresTranspose) Jacobian.BuildILUPreconditioner();
         break;
       case JACOBI:
-        Jacobian.BuildJacobiPreconditioner(RequiresTranspose);
+      case LINELET:
+        if (RequiresTranspose) Jacobian.BuildJacobiPreconditioner();
+        break;
+      case LU_SGS:
+        /*--- Nothing to build. ---*/
         break;
       case PASTIX_ILU: case PASTIX_LU_P: case PASTIX_LDLT_P:
-        Jacobian.BuildPastixPreconditioner(geometry, config, KindPrecond, RequiresTranspose);
+        /*--- It was already built. ---*/
         break;
       default:
         SU2_MPI::Error("The specified preconditioner is not yet implemented for the discrete adjoint method.", CURRENT_FUNCTION);
         break;
     }
 
+    /*--- Start recording if it was stopped for the linear solver ---*/
+#ifdef CODI_REVERSE_TYPE
+    AD::StartRecording();
+    SU2_OMP_BARRIER
+
+    SU2_OMP_MASTER {
+      AD::SetExtFuncOut(&LinSysSol[0], LinSysSol.GetLocSize());
+      AD::FuncHelper->addUserData(&LinSysRes);
+      AD::FuncHelper->addUserData(&LinSysSol);
+      AD::FuncHelper->addUserData(&Jacobian);
+      AD::FuncHelper->addUserData(geometry);
+      AD::FuncHelper->addUserData(config);
+      AD::FuncHelper->addUserData(this);
+    }
+    END_SU2_OMP_MASTER
+    SU2_OMP_BARRIER
+
+    AD::FuncHelper->addToTape(CSysSolve_b<ScalarType>::Solve_b);
+    SU2_OMP_BARRIER
+
+    SU2_OMP_MASTER
     AD::EndExtFunc();
+    END_SU2_OMP_MASTER
+    SU2_OMP_BARRIER
```

Comment:
```suggestion
```
Maybe this last barrier is not needed? Since end just releases the FuncHelper.

Reply:
:+1: